File: SandboxMachine/src/aiexecutor.cpp
/*
 * aiexecutor.cpp
 *
 * Copyright (C) 2015 Dominic S. Meiser <meiserdo@web.de>
 *
 * This work is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or any later
 * version.
 *
 * This work is distributed in the hope that it will be useful, but without
 * any warranty; without even the implied warranty of merchantability or
 * fitness for a particular purpose. See version 2 and version 3 of the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "aiexecutor.h"
#include "logger.h"
#include "mirrorclient.h"
#include "workerclient.h"

#include <error.h>
#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#include <QCoreApplication>
#include <QRegularExpression>
#include <QTemporaryDir>

#define MAX_BOX_ID 100
#define MAX_BOX_ID_RETRIES 1000

AiExecutor::AiExecutor (int id, int version, const QString &lang, const QUuid &uuid)
	: _id(id)
	, _version(version)
	, _lang(lang)
	, _uuid(uuid)
{
	// eine freie id für isolate finden
	for (int i = 0; i < MAX_BOX_ID_RETRIES; i++)
	{
		printf("loop nr. %d\n", i);
		QThread::sleep(1);
		boxid = rand() % MAX_BOX_ID;
		if (!QDir("/tmp/box/" + QString::number(boxid)).exists())
			break;
	}
	
	// isolate initialisieren
	QString cmd = "isolate --cg --init -b " + QString::number(boxid);
	printf("$ %s\n", qPrintable(cmd));
	FILE *isolateInit = popen(qPrintable(cmd), "r");
	if (!isolateInit)
	{
		fprintf(stderr, "Fehler beim initialisieren von isolate\n");
		abort = true;
		return;
	}
	QTextStream isolateInitIn(isolateInit);
	boxdir = isolateInitIn.readAll().trimmed();
	boxdir.mkdir("box");
	dir = boxdir.absoluteFilePath("box/");
}

AiExecutor::~AiExecutor()
{
	disconnect(&proc, SIGNAL(finished(int)), this, SLOT(cleanup(int)));
	if (start)
		delete start;
}

void AiExecutor::runAi ()
{
	if (abort)
	{
		LOG_CRITICAL << "Weigere mich aufgrund vorheriger Fehler die KI zu laden und zu starten";
		emit finished(uuid());
		return;
	}
	connect(this, SIGNAL(startAi()), this, SLOT(download()));
	connect(this, SIGNAL(downloaded()), this, SLOT(generateProps()));
	connect(this, SIGNAL(propsGenerated()), this, SLOT(executeAi()));
	connect(QCoreApplication::instance(), SIGNAL(aboutToQuit()), this, SLOT(terminateAi()));
	emit startAi();
}

void AiExecutor::download ()
{
	if (abort)
	{
		LOG_CRITICAL << "Weigere mich aufgrund vorheriger Fehler die KI zu herunterzuladen";
		emit downloaded();
		return;
	}
	
	// das Archiv über den Mirror des Workers herunterladen
	MirrorClient mirror(config->value("Worker/Host").toString(), config->value("Worker/MirrorPort").toUInt());
	if (!mirror.waitForConnected())
	{
		LOG_CRITICAL << "Fehler: Konnte nicht mit dem Mirror verbinden";
		abort = true;
		emit downloaded();
		return;
	}
	binArchive = dir.absoluteFilePath("bin.tar.bz2");
	if (!mirror.retrieveAi(id(), version(), binArchive))
	{
		LOG_CRITICAL << "Fehler: Konnte KI nicht über den Mirror des Workers herunterladen";
		abort = true;
		emit downloaded();
		return;
	}
	
	// die Privilegien der Archiv-Datei anpassen
	if (chown(qPrintable(binArchive), 0, 0) != 0)
	{
		perror("Fehler: Kann den Benutzer von der binArchiv-Datei nicht ändern");
		fprintf(stderr, "        Datei: %s\n", qPrintable(binArchive));
		abort = true;
		emit downloaded();
		return;
	}
	if (chmod(qPrintable(binArchive), S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH) != 0)
	{
		perror("Fehler: Kann die Dateiprivilegien nicht auf drw-r--r-- setzen");
		fprintf(stderr, "        Datei: %s\n", qPrintable(binArchive));
		abort = true;
		emit downloaded();
		return;
	}
	
	// das bin-Verzeichnis anlegen
	dir.mkdir("bin");
	binDir = dir.absoluteFilePath("bin");
	if (chown(qPrintable(binDir.absolutePath()), 0, 0) != 0)
	{
		perror("Fehler: Kann den Benutzer vom KI Verzeichnis nicht ändern");
		fprintf(stderr, "        Verzeichnis: %s\n", qPrintable(binDir.absolutePath()));
		abort = true;
		emit downloaded();
		return;
	}
	
	// tar ausführen
	QString cmd = "bsdtar xfj " + dir.absoluteFilePath("bin.tar.bz2") + " --uid 0 --gid 0 -C " + binDir.absolutePath();
	printf("$ %s\n", qPrintable(cmd));
	if (system(qPrintable(cmd)) != 0)
	{
		abort = true;
		emit downloaded();
		return;
	}
	
	// die start.ini laden
	start = new QSettings(binDir.absoluteFilePath("start.ini"), QSettings::IniFormat);
	LOG_DEBUG << "Libraries: " + start->value("Libraries").toString();
	for (uint i = 0; i < start->value("Libraries").toUInt(); i++)
	{
		MirrorClient mirror0(config->value("Worker/Host").toString(), config->value("Worker/MirrorPort").toUInt());
		if (!mirror0.waitForConnected())
		{
			LOG_CRITICAL << "Fehler: Konnte nicht mit dem Mirror verbinden";
			abort = true;
			emit downloaded();
			return;
		}
		
		QString path = binDir.absoluteFilePath(start->value("Lib" + QString::number(i) + "/Path").toString());
		binDir.mkdir(path);
		if (chown(qPrintable(path), 0, 0) != 0)
		{
			perror("Fehler: Kann den Benutzer vom KI Verzeichnis nicht ändern");
			fprintf(stderr, "        Verzeichnis: %s\n", qPrintable(path));
			abort = true;
			emit downloaded();
			return;
		}
		
		if (!mirror0.retrieveLib(start->value("Language").toString(), start->value("Lib" + QString::number(i) + "/Name").toString(), path + "/.tar.bz2"))
		{
			LOG_CRITICAL << "Fehler: Konnte Bibliothek nicht über den Mirror des Workers herunterladen";
			abort = true;
			emit downloaded();
			return;
		}
		
		// tar ausführen
		// wenn ich das nicht als root ausführe kriegt das automatisch die richtigen owner
		QString cmd = "bsdtar xfj \"" + path + "/.tar.bz2\" --uid 0 --gid 0 -C \"" + path + "/\"";
		printf("$ %s\n", qPrintable(cmd));
		if (system(qPrintable(cmd)) != 0)
		{
			abort = true;
			emit downloaded();
			return;
		}
		
		if (chmod(qPrintable(path), S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH) != 0)
		{
			perror("Fehler: Kann die Verzeichnisprivilegien nicht auf drwxr-xr-x setzen");
			fprintf(stderr, "        Verzeichnis: %s\n", qPrintable(path));
			abort = true;
			emit downloaded();
			return;
		}
	}
	
	// die Privilegien für das bin-Verzeichnis anpassen
	if (chmod(qPrintable(binDir.absolutePath()), S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH) != 0)
	{
		perror("Fehler: Kann die Verzeichnisprivilegien nicht auf drwxr-xr-x setzen");
		fprintf(stderr, "        Verzeichnis: %s\n", qPrintable(binDir.absolutePath()));
		abort = true;
		emit downloaded();
		return;
	}
	
	emit downloaded();
}

void AiExecutor::generateProps ()
{
	if (abort)
	{
		LOG_CRITICAL << "Weigere mich aufgrund vorheriger Fehler die KI-Properties zu erstellen";
		emit propsGenerated();
		return;
	}
	
	aiProp = dir.absoluteFilePath("ai.prop");
	QFile file(aiProp);
	if (!file.open(QIODevice::WriteOnly))
	{
		LOG_CRITICAL << "Kann die KI Properties nicht beschreiben (" + aiProp + " ): " + file.errorString();
		abort = true;
		emit propsGenerated();
		return;
	}
	file.write("# GENERATED FILE - DO NOT EDIT\n");
	
	configMutex->lock();
	config->beginGroup("Worker");
	file.write("turnierserver.worker.host=" + config->value("Host").toByteArray() + "\n");
	file.write("turnierserver.worker.server.port=" + config->value("Port").toByteArray() + "\n");
	file.write("turnierserver.worker.server.aichar=A\n"); // vlt sollte das in die config der sandbox
	file.write("turnierserver.serializer.compress.worker=" + config->value("AiSerializerCompress").toByteArray() + "\n");
	file.write("turnierserver.ai.uuid=" + uuid().toString().replace(QRegularExpression("[\\{\\}]"), "").toUtf8() + "\n");
	config->endGroup();
	configMutex->unlock();
	
	file.close();
	
	// die Privilegien anpassen
	if (chown(qPrintable(aiProp), 0, 0) != 0)
	{
		perror("Fehler: Kann den Benutzer von der KI Properties-Datei nicht ändern");
		fprintf(stderr, "        Datei: %s\n", qPrintable(aiProp));
		abort = true;
		emit propsGenerated();
		return;
	}
	if (chmod(qPrintable(aiProp), S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH) != 0)
	{
		perror("Fehler: Kann die Dateiprivilegien nicht auf drw-r--r-- setzen");
		fprintf(stderr, "        Datei: %s\n", qPrintable(aiProp));
		abort = true;
		emit propsGenerated();
		return;
	}
	
	emit propsGenerated();
}

void AiExecutor::executeAi ()
{
	if (abort)
	{
		LOG_CRITICAL << "Weigere mich aufgrund vorheriger Fehler die KI zu starten";
		worker->sendMessage(uuid(), 'T'); // T weil interner Fehler
		emit finished(uuid());
		return;
	}
	
	proc.setProgram("isolate");
	LOG_DEBUG << "Habe das Programm auf sandboxd_helper gesetzt";
	QString cmd = commands.value(lang());
	if (cmd.isEmpty())
		cmd = start->value("Command").toString();
	LOG_DEBUG << "Der cmd ist " + cmd;
	QStringList args;
	args << "--cg" << "-p";
	args << "--dir=/etc/=" + etcPath;
	args << "--dir=/usr/lib/jvm";
	args << "--run" << "-b" << QString::number(boxid) << "--" << cmd;
	args << start->value("Arguments").toStringList();
	args << aiProp;
	LOG_DEBUG << "Der Befehl ist " + cmd;
	//QStringList args;
	proc.setArguments(args);
	LOG_DEBUG << "Habe die Argumente gesetzt";
	qDebug() << "$" << proc.program() << proc.arguments();
	printf("$ %s %s", qPrintable(proc.program()), qPrintable(QVariant(proc.arguments()).toString()));
	proc.setProcessChannelMode(QProcess::ForwardedChannels);
	//QThread::sleep(1);
	//proc.start("id", args);
	proc.start();
	if (!proc.waitForStarted(timeout()))
	{
		LOG_CRITICAL << "Die KI ist nicht gestartet";
		abort = true;
		worker->sendMessage(uuid(), 'T');
		emit finished(uuid());
		return;
	}
	connect(&proc, SIGNAL(finished(int)), this, SLOT(cleanup(int)));
	
	worker->sendMessage(uuid(), 'S');
}

void AiExecutor::cleanup (int retCode)
{
	LOG_DEBUG << "die KI hat sich mit dem Statuscode " + QString::number(retCode) + " beendet";
	worker->sendMessage(uuid(), 'F');
	emit finished(uuid());
	QString cmd = "isolate --cleanup -b " + QString::number(boxid);
	printf("$ %s\n", qPrintable(cmd));
	system(qPrintable(cmd));
}

void AiExecutor::terminateAi ()
{
	LOG_INFO << "AiExecutor::terminateAi() called";
	//disconnect(&proc, SIGNAL(finished(int)), this, SLOT(cleanup(int)));
	proc.kill();
	if (!proc.waitForFinished())
	{
		perror("Fehler beim Töten der KI");
	}
	worker->sendMessage(uuid(), 'T');
	emit finished(uuid());
}

void AiExecutor::killAi ()
{
	LOG_INFO << "AiExecutor::killAi() called";
	//disconnect(&proc, SIGNAL(finished(int)), this, SLOT(cleanup(int)));
	proc.kill();
	if (!proc.waitForFinished())
	{
		perror("Fehler beim Töten der KI");
	}
	worker->sendMessage(uuid(), 'K');
	emit finished(uuid());
}

File: SandboxManager/src/main/java/org/pixelgaffer/turnierserver/sandboxmanager/AiExecutor.java
/*
 * AiExecutor.java
 *
 * Copyright (C) 2015 Pixelgaffer
 *
 * This work is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or any later
 * version.
 *
 * This work is distributed in the hope that it will be useful, but without
 * any warranty; without even the implied warranty of merchantability or
 * fitness for a particular purpose. See version 2 and version 3 of the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.pixelgaffer.turnierserver.sandboxmanager;

import static java.lang.Math.min;
import static org.pixelgaffer.turnierserver.FileOwnerChanger.changeOwner;
import static org.pixelgaffer.turnierserver.PropertyUtils.getString;
import static org.pixelgaffer.turnierserver.PropertyUtils.getStringRequired;
import static org.pixelgaffer.turnierserver.sandboxmanager.SandboxMain.commands;
import static org.pixelgaffer.turnierserver.sandboxmanager.SandboxMain.etc;
import java.io.BufferedReader;
import java.io.EOFException;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.lang.ProcessBuilder.Redirect;
import java.security.NoSuchAlgorithmException;
import java.util.LinkedList;
import java.util.List;
import java.util.Properties;
import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
import org.apache.commons.compress.compressors.bzip2.BZip2CompressorInputStream;
import org.pixelgaffer.turnierserver.Airbrake;
import lombok.Getter;

public class AiExecutor implements Runnable
{
	@SuppressWarnings("serial")
	public class AiStartException extends Exception
	{
		public AiStartException ()
		{
			super();
		}
		
		public AiStartException (String message, Throwable cause, boolean enableSuppression, boolean writableStackTrace)
		{
			super(message, cause, enableSuppression, writableStackTrace);
		}
		
		public AiStartException (String message, Throwable cause)
		{
			super(message, cause);
		}
		
		public AiStartException (String message)
		{
			super(message);
		}
		
		public AiStartException (Throwable cause)
		{
			super(cause);
		}
	}
	
	@Getter
	private Job job;
	@Getter
	private JobControl jobControl;
	
	@Getter
	private int boxid;
	private String boxdir;
	private File dir, binArchive, binDir, aiProp;
	private Properties start;
	private Process proc;
	
	public AiExecutor (Job job, JobControl ctrl) throws AiStartException
	{
		this.job = job;
		jobControl = ctrl;
		
		// isolate id finden
		boxid = (getJob().getId() + 100) % 100;
		
		// isolate initialisieren
		try
		{
			ProcessBuilder pb = new ProcessBuilder("isolate", "--cg", "--init", "-b", Integer.toString(boxid));
			pb.redirectError(Redirect.INHERIT);
			System.out.println("$ " + pb.command());
			Process p = pb.start();
			BufferedReader in = new BufferedReader(new InputStreamReader(p.getInputStream()));
			boxdir = in.readLine();
			if (p.waitFor() != 0)
				throw new AiStartException("Error while initialising isolate (exit code: " + p.exitValue() + ")");
			dir = new File(boxdir, "box");
			dir.mkdirs();
		}
		catch (AiStartException e)
		{
			throw e;
		}
		catch (Exception e)
		{
			throw new AiStartException(e);
		}
	}
	
	@Override
	public void run ()
	{
		try
		{
//			if (getJob().getId() < 0)
//				throw new AiStartException("Simulierter Crash");
			download();
			generateProps();
			executeAi();
		}
		catch (Throwable e)
		{
			Airbrake.log(e).printStackTrace();
			jobControl.jobFinished(getJob().getUuid());
			SandboxMain.getClient().sendMessage(getJob().getUuid(), 'T');
		}
	}
	
	private static void makeReadOnly (File f, boolean executable) throws IOException
	{
		changeOwner(f, "root");
		f.setExecutable(executable);
		f.setReadable(true);
		f.setWritable(true, true);
	}
	
	private static void extract (File f, File d) throws IOException
	{
		System.out.println("extract(" + f + ", " + d + ")");
		TarArchiveInputStream in = new TarArchiveInputStream(new BZip2CompressorInputStream(new FileInputStream(f)));
		TarArchiveEntry entry;
		while ((entry = in.getNextTarEntry()) != null)
		{
			if (entry.isDirectory())
				new File(d, entry.getName()).mkdir();
			else
			{
				File file = new File(d, entry.getName());
				System.out.println(file.getAbsolutePath());
				OutputStream out = new FileOutputStream(file);
				long size = entry.getSize();
				long read = 0;
				while (size > read)
				{
					int toRead = (int)min(size - read, Integer.MAX_VALUE);
					while (toRead > 0)
					{
						byte buf[] = new byte[min(toRead, 8192)];
						int r = in.read(buf);
						if (r <= 0)
						{
							out.close();
							in.close();
							throw new EOFException();
						}
						toRead -= r;
						read += r;
						out.write(buf, 0, r);
					}
				}
				out.close();
			}
		}
		in.close();
	}
	
	protected void download () throws NoSuchAlgorithmException, IOException
	{
		binArchive = new File(dir, "bin.tar.bz2");
		MirrorClient.retrieveAi(getJob().getId(), getJob().getVersion(), binArchive.getAbsolutePath());
		makeReadOnly(binArchive, false);
		binDir = new File(dir, "bin");
		binDir.mkdir();
		extract(binArchive, binDir);
		
		start = new Properties();
		start.load(new FileInputStream(new File(binDir, "start.prop")));
		int libs = Integer.parseInt(start.getProperty("libraries.size"));
		SandboxMain.getLogger().debug("Libraries: " + libs);
		for (int i = 0; i < libs; i++)
		{
			File path = new File(binDir, start.getProperty("libraries." + i + ".path"));
			path.mkdirs();
			MirrorClient.retrieveLib(getJob().getLang(), start.getProperty("libraries." + i + ".name"),
					new File(path, ".tar.bz2").getAbsolutePath());
			makeReadOnly(new File(path, ".tar.bz2"), false);
			extract(new File(path, ".tar.bz2"), path);
		}
	}
	
	protected void generateProps () throws IOException
	{
		aiProp = new File(dir, "ai.prop");
		Properties aiProps = new Properties();
		aiProps.put("turnierserver.worker.host", getStringRequired("worker.host"));
		aiProps.put("turnierserver.worker.server.port", getStringRequired("worker.port"));
		aiProps.put("turnierserver.worker.server.aichar", "A");
		aiProps.put("turnierserver.serializer.compress.worker", getStringRequired("turnierserver.serializer.compress.worker"));
		aiProps.put("turnierserver.ai.uuid", getJob().getUuid().toString());
		aiProps.put("turnierserver.debug", getString("turnierserver.debug", "false"));
		aiProps.store(new FileOutputStream(aiProp), "GENERATED FILE - DO NOT EDIT");
		makeReadOnly(aiProp, false);
	}
	
	protected void executeAi () throws IOException
	{
		List<String> cmd = new LinkedList<>();
		cmd.add("isolate");
		String command;
		if (commands.get(getJob().getLang()).isEmpty())
			command = start.getProperty("command");
		else
			command = commands.get(getJob().getLang());
		if (command.startsWith("."))
		{
			SandboxMain.getLogger()
					.debug("Flagging " + new File(binDir, command.substring(2)).getAbsolutePath() + " as executable");
			new File(binDir, command.substring(2)).setExecutable(true);
			command = "/box/bin" + command.substring(1);
		}
		cmd.add("--cg");
		cmd.add("-p");
		cmd.add("--share-net");
		cmd.add("-q");
		cmd.add("0,0");
		cmd.add("--time=" + job.getTimeout());
		cmd.add("--dir=/etc/=" + etc.getAbsolutePath());
		cmd.add("--dir=/usr/lib/jvm/");
		cmd.add("-c");
		cmd.add("/box/bin/");
		cmd.add("--env=LANG");
		for (int i = 0; i < Integer.parseInt(start.getProperty("environment.size")); i++)
		{
			cmd.add("--env=" + start.getProperty("environment." + i + ".key") + "="
					+ start.getProperty("environment." + i + ".value"));
		}
		cmd.add("--run");
		cmd.add("-b");
		cmd.add(Integer.toString(boxid));
		cmd.add("--");
		cmd.add(command);
		for (int i = 0; i < Integer.parseInt(start.getProperty("arguments.size")); i++)
		{
			cmd.add(start.getProperty("arguments." + i));
		}
		cmd.add("/box/" + aiProp.getName());
		SandboxMain.getLogger().debug("Der Befehl ist " + cmd);
		
		ProcessBuilder pb = new ProcessBuilder(cmd);
		pb.redirectErrorStream(true);
		pb.redirectOutput(Redirect.INHERIT);
		proc = pb.start();
		SandboxMain.getClient().sendMessage(getJob().getUuid(), 'S');
		
		new Thread( () -> {
			int ret;
			try
			{
				ret = proc.waitFor();
				SandboxMain.getLogger().debug("Die KI hat sich mit dem Statuscode " + ret + " beendet");
				SandboxMain.getClient().sendMessage(getJob().getUuid(), 'F');
				ProcessBuilder pb0 = new ProcessBuilder("isolate", "--cleanup", "-b", Integer.toString(boxid));
				SandboxMain.getLogger().debug(pb0.command());
				pb0.redirectErrorStream(true);
				pb0.redirectOutput(Redirect.INHERIT);
				if (pb0.start().waitFor() != 0)
					SandboxMain.getLogger().critical("Fehler beim Aufräumen von isolate");
				jobControl.jobFinished(getJob().getUuid());
			}
			catch (Exception e)
			{
				Airbrake.log(e).printStackTrace();
			}
		} , "IsolateCleanup").start();
	}
	
	public void terminateAi ()
	{
		SandboxMain.getLogger().info("terminiere " + getJob());
		proc.destroy();
		SandboxMain.getClient().sendMessage(getJob().getUuid(), 'T');
	}
	
	public void killAi ()
	{
		SandboxMain.getLogger().info("töte " + getJob());
		proc.destroy();
		SandboxMain.getClient().sendMessage(getJob().getUuid(), 'K');
	}
}
